### PR TITLE
Add ErrorHandler DLQ API to Python

### PR DESCRIPTION
**Pull request by robertwb**
It recently came to my attention that this was only added to Java.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @johnjcasey

**Review by Abacn — COMMENTED** [review]
while pending @johnjcasey review also went through once and mostly LGTM

File: sdks/python/apache_beam/transforms/error_handling.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+
+class ErrorHandler:
+  """ErrorHandlers are used to skip and otherwise process bad records.
+
+  Error handlers allow one to implement the "dead letter queue" pattern in
+  a fluent manner, disaggregating the error processing specification from
+  the main processing chain.
+
+  They is typically used as follows::
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/error_handling.py]: Done.

**Comment by robertwb** [comment]
Thanks. Any other thoughts on this, @johnjcasey ?

**Comment by robertwb** [comment]
Another ping on this.

**Comment by Abacn** [comment]
would be nice to get this in by next release. For past contributor/reviewers on this topic, maybe @bzablocki (who reviewed #27145) could take another look if needed?

File: sdks/python/apache_beam/transforms/core.py [hunk 2]
```diff
@@ -1633,7 +1636,19 @@ def with_exception_handling(
         use_subprocess,
         threshold,
         threshold_windowing,
-        timeout)
+        timeout,
+        error_handler)
+
+  def with_error_handler(self, error_handler, **exception_handling_kwargs):
```
Comment (bzablocki) [inline sdks/python/apache_beam/transforms/core.py:1642]: Could we also have a unit test that would show how the `**exception_handling_kwargs` can be used?

**Review by bzablocki — APPROVED** [review]
Small comments, LGTM overall

File: sdks/python/apache_beam/transforms/error_handling_test.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      error_pcoll = error_handler.output()
+
+      assert_that(result, equal_to(['A', 'Bb']), label='CheckGood')
+      assert_that(error_pcoll, equal_to(['error: cccc']), label='CheckBad')
+
+  def test_error_handling_pardo(self):
+    with beam.Pipeline() as p:
+      pcoll = p | beam.Create(['a', 'bb', 'cccc'])
+      with error_handling.ErrorHandler(
+          beam.Map(lambda x: "error: %s" % x[0])) as error_handler:
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/error_handling_test.py:81]: `PTransformWithErrors.with_error_handling()` returns a PCollection of elements as errors, whereas the standard `ParDo(...).with_error_handling()` returns the Python equivalent of bad records that attach the bad elements to the exception thrown.

File: sdks/python/apache_beam/transforms/error_handling_test.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      pcoll = p | beam.Create(['a', 'bb', 'cccc'])
+      with error_handling.CollectingErrorHandler() as error_handler:
+        result = pcoll | beam.Map(
+            exception_throwing_map, limit=3).with_error_handler(error_handler)
+      error_pcoll = error_handler.output() | beam.Map(lambda x: x[0])
+
+      assert_that(result, equal_to(['A', 'Bb']), label='CheckGood')
+      assert_that(error_pcoll, equal_to(['cccc']), label='CheckBad')
+
+  def test_error_on_collecting_error_handler_without_output_retrieval(self):
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/error_handling_test.py:108]: This is closed (due to the context) but not consumed.

File: sdks/python/apache_beam/transforms/core.py (hunk 2 above)
Comment (robertwb) [reply sdks/python/apache_beam/transforms/core.py:1642]: Good call. Done.

**Review by robertwb — COMMENTED** [review]
Thanks for the review!

**Pull request merged**